### PR TITLE
Dev daniel

### PR DESCRIPTION
**Pull request by danielbattistini**
Added Lc config files, minor changes in optimisation/ScanSelectionsTree.py and its config files.

### CODE REVIEW

File: ML/config_Training_LctopK0s_13TeV_pt_4_6.yml [hunk 1]
```diff
@@ -0,0 +1,154 @@
+input: # files to use, set FD to null for binary classification
```
Comment (fgrosa) [inline ML/config_Training_LctopK0s_13TeV_pt_4_6.yml:1]: Is this file needed since you already have the one with 4-50 pt interval?

File: ML/config_Application_LctopK0s_13TeV_pt_4_50.yml [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        [
+            ~/DmesonAnalysis/ML/trained_models/LctopK0s/13TeV/pt4_6/ModelHandler_pT_4_6.pickle,
+            ~/DmesonAnalysis/ML/trained_models/LctopK0s/13TeV/pt6_50/ModelHandler_pT_6_50.pickle,
+        ]
+        # list of saved ModelHandler (path+file), compatible with the pt bins
+
+plots:
+    plotting_columns:
+        [
+            inv_mass,
```
Comment (fgrosa) [inline ML/config_Application_LctopK0s_13TeV_pt_4_50.yml:105]: here you could include all the variables that you use in the training

File: ML/config_Training_LctopK0s_13TeV_pt_2_4.yml [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    min: [2] # list
+    max: [4] # list
+
+data_prep:
+    filt_bkg_mass: inv_mass < 2.15 or inv_mass > 2.4 # pandas query to select bkg candidates
+    dataset_opt:
+        "max_signal" # change how the dataset is built, options available: 'equal', 'max_signal'
+        # 'equal' -> same number of prompt/FD/bkg (not using all the signal available)
+        # 'max_signal' -> try to use all the signal (prompt and FD) + add n_bkg = bkg_mult * (n_prompt + n_FD)
+    bkg_mult: [0.08] # list of multipliers for (nPrompt + nFD) used to determine nCandBkg in the 'max_signal' option
```
Comment (fgrosa) [inline ML/config_Training_LctopK0s_13TeV_pt_2_4.yml:38]: why did you set an amount of background which is only 0.08 times the signal? Maybe it would be better to use the same amount of signal and background at low pT where we don't have problems with the statistics

File: ML/config_Training_LctopK0s_13TeV_pt_2_4.yml [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+            ~/DmesonAnalysis/ML/trained_models/LctopK0s/ModelHandler_pT_2_4.pickle,
+            # ~/DmesonAnalysis/ML/trained_models/LctopK0s/ModelHandler_pT_4_6.pickle,
+            # ~/DmesonAnalysis/ML/trained_models/LctopK0s/ModelHandler_pT_6_50.pickle,
+        ]
+        # list of saved ModelHandler (path+file), compatible with the pt bins
+
+plots:
+    plotting_columns:
+        [
+            inv_mass,
```
Comment (fgrosa) [inline ML/config_Training_LctopK0s_13TeV_pt_2_4.yml:147]: same as above

File: ML/config_Training_LctopK0s_13TeV_pt_4_50.yml [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    min: [4, 6] # list
+    max: [6, 50] # list
+
+data_prep:
+    filt_bkg_mass: inv_mass < 2.15 or inv_mass > 2.4 # pandas query to select bkg candidates
+    dataset_opt:
+        "max_signal" # change how the dataset is built, options available: 'equal', 'max_signal'
+        # 'equal' -> same number of prompt/FD/bkg (not using all the signal available)
+        # 'max_signal' -> try to use all the signal (prompt and FD) + add n_bkg = bkg_mult * (n_prompt + n_FD)
+    bkg_mult: [0.08, 0.08] # list of multipliers for (nPrompt + nFD) used to determine nCandBkg in the 'max_signal' option
```
Comment (fgrosa) [inline ML/config_Training_LctopK0s_13TeV_pt_4_50.yml:27]: same as above: why did you set an amount of background which is only 0.08 times the signal? Is it because otherwise the fraction of background used is too high?

File: ML/config_Training_LctopK0s_13TeV_pt_4_50.yml [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        [
+            ~/DmesonAnalysis/ML/trained_models/LctopK0s/13TeV/pt4_6/ModelHandler_pT_4_6.pickle,
+            ~/DmesonAnalysis/ML/trained_models/LctopK0s/13TeV/pt6_50/ModelHandler_pT_6_50.pickle,
+        ]
+        # list of saved ModelHandler (path+file), compatible with the pt bins
+
+plots:
+    plotting_columns:
+        [
+            inv_mass,
```
Comment (fgrosa) [inline ML/config_Training_LctopK0s_13TeV_pt_4_50.yml:105]: same as above

File: configfiles/config_LctopK0s_data_tree.yml [hunk 7]
```diff
@@ -0,0 +1,17 @@
+isMC: no
+filename: ~/DmesonAnalysis/AnalysisResults/13TeV/data/AnalysisResults_LcpK0s_2016_kl.root
```
Comment (fgrosa) [inline configfiles/config_LctopK0s_data_tree.yml:2]: Is it intended to use only one period instead of the full sample (namely 2016, 2017, and 2018)?

File: configfiles/fit/config_Lc_Fit_pp13TeV.yml [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    MassMin: [1.78, 1.78]
+    MassMax: [2.10, 2.10]
+    Rebin: [6, 8]
+    InclSecPeak: [1, 1, 1, 1]
+    SigmaSecPeak: [0.008, 0.008, 0.008, 0.008]
+    SigmaFileSecPeak: ""
+    SigmaMultFactorSecPeak: 1.
+    FixSigmaToFirstPeak: 0
+    UseLikelihood: 1
+    BkgFunc: ["kExpo", "kExpo"]
```
Comment (fgrosa) [inline configfiles/fit/config_Lc_Fit_pp13TeV.yml:19]: Not relevant now, but you will see that for fitting the Lc invariant-mass spectrum a polynomial of second order (`Pol2`) would be better

File: configfiles/fit/config_Lc_Fit_pp13TeV.yml [hunk 9]
```diff
@@ -0,0 +1,20 @@
+pp5TeVFD:
+    #Meson: 'Lc'
```
Comment (fgrosa) [inline configfiles/fit/config_Lc_Fit_pp13TeV.yml:2]: This is needed when you will fit

File: configfiles/fit/config_Lc_Fit_pp13TeV.yml [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+    FixSigma: 1
+    SigmaFile: '../../Analyses/pp5TeV/Ds_wML_mult/outputs/100320/raw_yield/RawYield_Ds_data_norm_pt2_12.root'
```
Comment (fgrosa) [inline configfiles/fit/config_Lc_Fit_pp13TeV.yml:4]: you can remove this one and leave `''`

File: configfiles/fit/config_Lc_Fit_pp13TeV.yml [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+    SigmaMultFactor: 1. #float or MinusUnc, PlusUnc
+    FixMean: 0
+    MeanFile: ""
+    PtMin: [4, 6]
+    PtMax: [6, 12]
+    MassMin: [1.78, 1.78]
+    MassMax: [2.10, 2.10]
+    Rebin: [6, 8]
+    InclSecPeak: [1, 1, 1, 1]
```
Comment (fgrosa) [inline configfiles/fit/config_Lc_Fit_pp13TeV.yml:13]: this should be all 0 for each pT bin for the Lc (it should be 1 only for the Ds)

**Review by fgrosa — COMMENTED** [review]
Thanks for the work! In general I was just wondering whether the config files could be committed in a second time, when we will be able to use the new workstation and hence analyse all the pT bins together, but it is also fine to change them once we will have it. I then added some specific comments to your config files (more related to the analysis then the code)

File: ML/config_Training_LctopK0s_13TeV_pt_4_6.yml (hunk 1 above)
Comment (danielbattistini) [reply ML/config_Training_LctopK0s_13TeV_pt_4_6.yml:1]: No, I can remove it.

File: configfiles/config_LctopK0s_data_tree.yml (hunk 7 above)
Comment (danielbattistini) [reply configfiles/config_LctopK0s_data_tree.yml:2]: No, but this file is not used in the analysis yet. Anyways I can remove it you prefer, otherwise, I will upload the updated version when it's time.

**Review by fgrosa — APPROVED** [review]

File: optimisation/ScanSelectionsTree.py [hunk 12]
```diff
@@ -192,6 +191,9 @@
 cSignifVsRest, hSignifVsRest, cEstimVsCut, hEstimVsCut = [], [], [], []
 counter = 0
 for iPt, (ptMin, ptMax) in enumerate(zip(ptMins, ptMaxs)):
+    # reshuffle bkg and take only a fraction of it, seed fixed for reproducibility
+    dfBkg = dfBkg.sample(frac=fractionstokeep[iPt], random_state=42).reset_index(drop=True)
+
```
Comment (fcatalan92) [inline optimisation/ScanSelectionsTree.py:196]: Hi @danielbattistini, this looks like a possible bug to me. Here you are sampling and modifying the `bkg` dataframe integrated in pt, so from the second bin onward you are sampling from the "reduced" dataframe obtained before.
Moreover, but this was a problem probably already present, I think it is better to first select the pt range of your interest and then do the sampling. The difference is minor if your input file contains data from a small pt range but when you consider all the analysis pt range you could have problems.

Comment (danielbattistini) [reply optimisation/ScanSelectionsTree.py:196]: Thank you for notifying me! That's right, it should be fixed now.
